doc/adr/0012-modules-structure.md

```diff
@@ -0,0 +1,84 @@
+- `kernel` contains part of state channels (hylomorphism's idea),
+- `sdk` contains all the things shared between Global L0 and DAG L1, but not the $DAG itself,
+- `dag-shared` contains all the things shared between Global L0 and DAG L1 related to $DAG,
+- `core` is the Global L0 module (but also contains state channels that must be extracted to be reused in other modules),
+- `dag-l1` is the DAG L1 module ($DAG token Layer 1).
+
+```mermaid
+graph TD
+    A[Shared]
+    B[Keytool]
+    C[Kernel]
+    D[Wallet]
+    E[SDK]
+    F[DAG Shared]
+    G[Global L0]
+    H[DAG L1]
+    
+    B --> A
+    C --> A
+    D --> B
+    F --> E
+    E --> C
+    G --> F
+    H --> F
+```
+
+Furthermore, dependencies were partially explicitly declared and partially transient. After some time, it became unmaintainable.
+
+The second (and main) reason for restructuring modules is the fact that 3rd parties cannot use the SDK module in the current form. It does not expose all required operations, so internal modules were used directly, leaking the project's internal structure and locking further internal changes.
+
+## Decision
+
+`keytool` module will be splitted into: operations/functions that we will put to `shared` and exposing an app that will be combined into `wallet` module.
+Then, all the things related to node operations would be kept in `node-shared` module (including gossip protocol, consensus, etc.). This leads to emptying the `sdk` module, so we will be able to implement the `sdk` module as a proxy between the external world and the internal tessellation structure. `sdk` module should reexpose all necessary operations to be used by 3rd parties. For example - to create a keypair, global L0 can use the `shared` module directly, but 3rd party will use the `sdk` module that will reexpose keypair creation function. That way, the internal structure of tessellation will be hidden behind the facade.
+
+Modules' dependencies and libraries cannot be transient. We will use `% Optional` scope for all internal modules.
+
+- `shared` will contain base classes and extensions (like kryo or refined extensions), all the stuff used within the GlobalSnapshot (like Address, base Block or Transaction), base StateChannels types and key-related operations (like getting address from public key),
+- `wallet` will contain operations required by the wallet-cli (like creating transactions), but also will expose key operations (like creating a new key),
+- `dag-shared` will contain shared types and classes between Global L0 and DAG L1, related to $DAG (like $DAG block or $DAG transaction),
+- `node-shared` will contain all the operations required to launch a node (like base TessellationIO app, gossip protocol, cluster management or joining algorithm),
+- `global-l0` will be renamed from `core` and will remain the Global L0 module,
+- `dag-l1` will remain the DAG L1 module.
```

Comment:
Generally speaking I would rather see this list rephrased as definitions (in a present tense):

```
- `shared` - base library module containing `GlobalSnapshot` as well as extensions for 3rd party libraries
- `dag-shared` - an internal library module containing classes shared only between `dag-l1` and `global-l0`
- `node-shared` - an internal library module containing all the classes necessary to launch a node (like base TessellationIO app, gossip protocol, cluster management or joining algorithm),
- `sdk` - a public library module exposing a set of functionalities ... 
- `tools` an internal app used for generating transactions and state channel snapshots
- `wallet` - CLI app for creating transactions and managing keys
- `global-l0` - the Global L0 node app (renamed from `core`)
- `dag-l1` - the DAG L1 node app (remains unchanged)
```

Reply:
:+1: rephrased